File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-4.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        clones = { None : None }

        curr = head
        while curr:
            copy = Node(curr.val)
            clones[curr] = copy
            curr = curr.next
        
        curr = head
        while curr:
            copy = clones[curr]
            copy.next = clones[curr.next]
            copy.random = clones[curr.random]
            curr = curr.next
            
        return clones[head]

File: Data Structures & Algorithms/reverse-linked-list-ii/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode dummy = ListNode(-1, head);
        int n = right - left;

        // Get pre_swap
        ListNode* node_before_left = &dummy;
        for(int i = 0; i < left - 1; i++) {
            node_before_left = node_before_left->next;
        }

        ListNode* tail = node_before_left->next;
        ListNode* curr = tail;
        ListNode* prev = nullptr;
        while (n > -1) {
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
            n--;
        }

        tail->next = curr;
        node_before_left->next = prev;

        return dummy.next;
    }
};